### PR TITLE
Make Lua documentation generation more generic

### PR DESCRIPTION
**Pull request by asarium**
This adds JSON as an additional scripting documentation output format.
This can be used by external tools to examine the scripting
documentation and generate interface declarations or documentation for
any language. The existing documentation code is left largely in place.

The initial purpose of this is to allow using TypeScript with FSO by using
the [TypeScript to Lua](https://typescripttolua.github.io/) compiler.

### CODE REVIEW

**Comment by Goober5000** [comment]
I'm still skeptical of using `uint32_t` instead of `size_t`...

**Comment by asarium** [comment]
In these changes `size_t` would not be correct since the other variables are also unsigned ints...

File: qtfred/src/ui/widgets/sexp_tree.cpp [hunk 1]
```diff
@@ -5841,6 +5844,4 @@ void sexp_tree::deleteCurrentItem() {
 int sexp_tree::getCurrentItemIndex() const {
 	return item_index;
 }
-
-}
```
Comment (Goober5000) [inline qtfred/src/ui/widgets/sexp_tree.cpp]: Here's what's causing the compilation error.

**Comment by Goober5000** [comment]
Unfortunately, Travis is still complaining:

```
$ ./ci/travis/clang_tidy.sh

Running clang-tidy on changed files

../code/globalincs/flagset.h:4:10: error: 'bitset' file not found [clang-diagnostic-error]

#include <bitset>

         ^

../code/globalincs/vmallocator.h:6:10: error: 'deque' file not found [clang-diagnostic-error]

#include <deque>

         ^

../code/windows_stub/config.h:8:10: error: 'cstdio' file not found [clang-diagnostic-error]

#include <cstdio>

         ^

Errors or warnings were detected in the clang-tidy output!

The command "./ci/travis/clang_tidy.sh" exited with 1.
```

**Comment by Goober5000** [comment]
This looks promising:

https://stackoverflow.com/questions/26333823/clang-doesnt-see-basic-headers

**Comment by Goober5000** [comment]
No joy on that latest push. :(

```
$ ./ci/travis/clang_tidy.sh

Running clang-tidy on changed files

../code/globalincs/flagset.h:4:10: error: 'bitset' file not found [clang-diagnostic-error]

#include <bitset>

         ^

../code/globalincs/vmallocator.h:6:10: error: 'deque' file not found [clang-diagnostic-error]

#include <deque>

         ^

../code/windows_stub/config.h:8:10: error: 'cstdio' file not found [clang-diagnostic-error]

#include <cstdio>

         ^

Errors or warnings were detected in the clang-tidy output!

The command "./ci/travis/clang_tidy.sh" exited with 1.
```

**Comment by Goober5000** [comment]
Well now we're getting some different errors:
```
error: '/home/travis/build/scp-fs2open/fs2open.github.com/build/code/cotire/code_CXX_prefix.hxx' file not found [clang-diagnostic-error]
```
This is repeated numerous times.

Also:
```
/home/travis/build/scp-fs2open/fs2open.github.com/code/scripting/api/libs/options.cpp:32:37: warning: use nullptr [modernize-use-nullptr]

ADE_FUNC(persistChanges, l_Options, NULL,

                                    ^~~~

                                    nullptr
```

**Comment by asarium** [comment]
I was and am testing some things so for now you can ignore what Travis says. I'm taking care of it.

**Comment by Goober5000** [comment]
Ok.

**Pull request merged**